File: AgamCs/species_context.py
"""Exact-query species and topology-clade summaries.

The matching browser implementation lives in ``docs/assets/species-context.js``.
Both are tested against ``tests/fixtures/species-context-v1-cases.json``.
"""

from __future__ import annotations

import math
from collections.abc import Mapping, Sequence


ANALYSIS_VERSION = 'agamcs-species-context-v1'
SCHEMA_VERSION = 1
COORDINATE_CONVENTION = '1-based inclusive'
WINDOW_SIZE = 100
WINDOW_DETECTION_THRESHOLD = 0.8


def _tip_codes(node: str | Mapping) -> list[str]:
    if isinstance(node, str):
        return [node]
    children = node.get('children') if isinstance(node, Mapping) else None
    if not isinstance(children, list) or not children:
        raise ValueError('Topology nodes must be genome codes or non-empty child-bearing objects.')
    return [code for child in children for code in _tip_codes(child)]


def _clades(node: str | Mapping, path: tuple[str, ...] = ()) -> list[dict]:
    if isinstance(node, str):
        return []
    name = str(node.get('name') or '').strip()
    children = node.get('children')
    if not name or not isinstance(children, list) or not children:
        raise ValueError('Every topology clade must have a name and at least one child.')
    current_path = (*path, name)
    record = {
        'id': '/'.join(current_path),
        'name': name,
        'path': list(current_path),
        'member_codes': _tip_codes(node),
        'child_count': len(children),
        'is_polytomy': len(children) > 2,
    }
    return [record, *(item for child in children for item in _clades(child, current_path))]


def _longest_run(detected_by_position: Sequence[bool], start: int) -> dict:
    best_start = best_end = None
    run_start = None
    for offset, detected in enumerate([*detected_by_position, True]):
        if not detected and run_start is None:
            run_start = offset
        elif detected and run_start is not None:
            run_end = offset - 1
            if best_start is None or run_end - run_start > best_end - best_start:
                best_start, best_end = run_start, run_end
            run_start = None
    return {
        'start': start + best_start if best_start is not None else None,
        'end': start + best_end if best_end is not None else None,
        'bases': best_end - best_start + 1 if best_start is not None else 0,
    }


def _window(values_by_species: Sequence[Sequence[float]], chromosome: str,
            query_start: int, offset: int, length: int) -> dict:
    values = [
        float(value)
        for species_values in values_by_species
        for value in species_values[offset:offset + length]
        if value is not None and math.isfinite(float(value)) and float(value) != 0
    ]
    possible = len(values_by_species) * length
    return {
        'chromosome': chromosome,
        'start': query_start + offset,
        'end': query_start + offset + length - 1,
        'total_bases': length,
        'possible_species_bases': possible,
        'detected_bases': len(values),
        'detected_fraction': len(values) / possible,
        'mean_identity_detected': sum(values) / len(values) if values else None,
    }


def _summary(kind: str, identifier: str, name: str, member_codes: list[str],
             values_by_species: Sequence[Sequence[float]], chromosome: str,
             query_start: int, query_bases: int, **metadata) -> dict:
    detected_values = [
        float(value)
        for species_values in values_by_species
        for value in species_values
        if value is not None and math.isfinite(float(value)) and float(value) != 0
    ]
    detected_by_position = [
        any(
            value is not None and math.isfinite(float(value)) and float(value) != 0
            for value in (species_values[offset] for species_values in values_by_species)
        )
        for offset in range(query_bases)
    ]
    windows = [
        _window(values_by_species, chromosome, query_start, offset,
                min(WINDOW_SIZE, query_bases - offset))
        for offset in range(0, query_bases, WINDOW_SIZE)
    ]
    qualifying = [
        window for window in windows
        if window['total_bases'] == WINDOW_SIZE
        and window['detected_fraction'] >= WINDOW_DETECTION_THRESHOLD
        and window['mean_identity_detected'] is not None
    ]
    lowest = min(
        qualifying,
        key=lambda window: (window['mean_identity_detected'], window['start']),
        default=None,
    )
    possible = len(values_by_species) * query_bases
    return {
        'kind': kind,
        'id': identifier,
        'name': name,
        'member_codes': member_codes,
        'species_count': len(values_by_species),
        'query_bases': query_bases,
        'possible_species_bases': possible,
        'detected_bases': len(detected_values),
        'detected_fraction': len(detected_values) / possible,
        'mean_identity_detected': (
            sum(detected_values) / len(detected_values) if detected_values else None
        ),
        'longest_undetected_run': _longest_run(detected_by_position, query_start),
        'lowest_qualifying_identity_window': lowest,
        **metadata,
    }


def analyze_species_context(source_result: Mapping, topology: Mapping | None = None) -> dict:
    """Summarize exact CNEr stack values without treating zero as identity."""
    chromosome = str(source_result.get('chromosome') or '')
    start, end = source_result.get('start'), source_result.get('end')
    rows_value = source_result.get('stackRows')
    labels_value = source_result.get('stackSpecies')
    values = source_result.get('values')
    stack_value = values.get('stack') if isinstance(values, Mapping) else None
    rows = list(rows_value) if rows_value is not None else []
    labels = list(labels_value) if labels_value is not None else []
    stack = list(stack_value) if stack_value is not None else []
    topology = topology or source_result.get('stackTopology')
    if not chromosome or not isinstance(start, int) or not isinstance(end, int) or start < 1 or end < start:
        raise ValueError('Species-context analysis requires valid inclusive query coordinates.')
    if not rows or len(labels) != len(rows) or not isinstance(topology, Mapping):
        raise ValueError('Species-context analysis requires aligned species metadata and topology.')
    query_bases = end - start + 1
    if len(stack) != len(rows) * query_bases:
        raise ValueError('Species stack length must equal species rows multiplied by query bases.')
    tips = _tip_codes(topology.get('tree'))
    if tips != rows or len(set(tips)) != len(tips):
        raise ValueError('Species topology tips must uniquely match the stack-row order.')
    values_by_code = {
        code: stack[index * query_bases:(index + 1) * query_bases]
        for index, code in enumerate(rows)
    }
    species = [
        _summary('species', code, labels[index], [code], [values_by_code[code]],
                 chromosome, start, query_bases)
        for index, code in enumerate(rows)
    ]
    clades = []
    for clade in _clades(topology['tree']):
        member_codes = clade.pop('member_codes')
        identifier = clade.pop('id')
        name = clade.pop('name')
        clades.append(_summary(
            'clade', identifier, name, member_codes,
            [values_by_code[code] for code in member_codes],
            chromosome, start, query_bases, **clade,
        ))
    return {
        'schema_version': SCHEMA_VERSION,
        'analysis_version': ANALYSIS_VERSION,
        'coordinate_convention': COORDINATE_CONVENTION,
        'window_size': WINDOW_SIZE,
        'window_detection_threshold': WINDOW_DETECTION_THRESHOLD,
        'zero_semantics': 'No detected CNEr interval; not measured 0% identity.',
        'query': {'chromosome': chromosome, 'start': start, 'end': end, 'bases': query_bases},
        'species_count': len(rows),
        'species': species,
        'clades': clades,
    }

File: docs/assets/species-context.js
(function initialiseSpeciesContext(root, factory) {
  const api = factory();
  root.AgamCsSpeciesContext = api;
  if (typeof module === 'object' && module.exports) module.exports = api;
}(globalThis, () => {
  const ANALYSIS_VERSION = 'agamcs-species-context-v1';
  const SCHEMA_VERSION = 1;
  const COORDINATE_CONVENTION = '1-based inclusive';
  const WINDOW_SIZE = 100;
  const WINDOW_DETECTION_THRESHOLD = 0.8;

  function tipCodes(node) {
    if (typeof node === 'string') return [node];
    if (!node || !Array.isArray(node.children) || !node.children.length) throw new Error('Topology nodes must be genome codes or non-empty child-bearing objects.');
    return node.children.flatMap(tipCodes);
  }

  function cladeRecords(node, path = []) {
    if (typeof node === 'string') return [];
    const name = String(node?.name || '').trim();
    if (!name || !Array.isArray(node.children) || !node.children.length) throw new Error('Every topology clade must have a name and at least one child.');
    const currentPath = [...path, name];
    return [{
      id: currentPath.join('/'), name, path: currentPath,
      member_codes: tipCodes(node), child_count: node.children.length,
      is_polytomy: node.children.length > 2,
    }, ...node.children.flatMap((child) => cladeRecords(child, currentPath))];
  }

  function longestRun(detectedByPosition, start) {
    let bestStart = null; let bestEnd = null; let runStart = null;
    for (let offset = 0; offset <= detectedByPosition.length; offset += 1) {
      const detected = offset === detectedByPosition.length || Boolean(detectedByPosition[offset]);
      if (!detected && runStart == null) runStart = offset;
      else if (detected && runStart != null) {
        const runEnd = offset - 1;
        if (bestStart == null || runEnd - runStart > bestEnd - bestStart) [bestStart, bestEnd] = [runStart, runEnd];
        runStart = null;
      }
    }
    return { start: bestStart == null ? null : start + bestStart, end: bestEnd == null ? null : start + bestEnd, bases: bestStart == null ? 0 : bestEnd - bestStart + 1 };
  }

  function detectedValues(values) {
    return values.map(Number).filter((value) => Number.isFinite(value) && value !== 0);
  }

  function windowRecord(valuesBySpecies, chromosome, queryStart, offset, length) {
    let detected = 0; let identityTotal = 0;
    valuesBySpecies.forEach((valuesByPosition) => {
      for (let index = offset; index < offset + length; index += 1) {
        const value = Number(valuesByPosition[index]);
        if (Number.isFinite(value) && value !== 0) { detected += 1; identityTotal += value; }
      }
    });
    const possible = valuesBySpecies.length * length;
    return {
      chromosome, start: queryStart + offset, end: queryStart + offset + length - 1,
      total_bases: length, possible_species_bases: possible, detected_bases: detected,
      detected_fraction: detected / possible,
      mean_identity_detected: detected ? identityTotal / detected : null,
    };
  }

  function summary(kind, id, name, memberCodes, valuesBySpecies, chromosome, queryStart, queryBases, metadata = {}) {
    let detected = 0; let identityTotal = 0;
    const detectedByPosition = new Uint8Array(queryBases);
    valuesBySpecies.forEach((speciesValues) => {
      for (let offset = 0; offset < queryBases; offset += 1) {
        const value = Number(speciesValues[offset]);
        if (Number.isFinite(value) && value !== 0) {
          detected += 1; identityTotal += value; detectedByPosition[offset] = 1;
        }
      }
    });
    const windows = [];
    for (let offset = 0; offset < queryBases; offset += WINDOW_SIZE) windows.push(windowRecord(valuesBySpecies, chromosome, queryStart, offset, Math.min(WINDOW_SIZE, queryBases - offset)));
    const qualifying = windows.filter((window) => window.total_bases === WINDOW_SIZE && window.detected_fraction >= WINDOW_DETECTION_THRESHOLD && window.mean_identity_detected != null);
    qualifying.sort((left, right) => left.mean_identity_detected - right.mean_identity_detected || left.start - right.start);
    const possible = valuesBySpecies.length * queryBases;
    return {
      kind, id, name, member_codes: memberCodes, species_count: valuesBySpecies.length,
      query_bases: queryBases, possible_species_bases: possible, detected_bases: detected,
      detected_fraction: detected / possible,
      mean_identity_detected: detected ? identityTotal / detected : null,
      longest_undetected_run: longestRun(detectedByPosition, queryStart),
      lowest_qualifying_identity_window: qualifying[0] || null,
      ...metadata,
    };
  }

  function analyzeSpeciesContext(source, topology = null) {
    const chromosome = String(source?.chromosome || '');
    const { start, end } = source || {};
    const rows = [...(source?.stackRows || [])];
    const labels = [...(source?.stackSpecies || [])];
    const stack = source?.values?.stack;
    const selectedTopology = topology || source?.stackTopology;
    if (!chromosome || !Number.isSafeInteger(start) || !Number.isSafeInteger(end) || start < 1 || end < start) throw new Error('Species-context analysis requires valid inclusive query coordinates.');
    if (!rows.length || labels.length !== rows.length || !selectedTopology) throw new Error('Species-context analysis requires aligned species metadata and topology.');
    const queryBases = end - start + 1;
    if (!stack || stack.length !== rows.length * queryBases) throw new Error('Species stack length must equal species rows multiplied by query bases.');
    const tips = tipCodes(selectedTopology.tree);
    if (tips.length !== rows.length || new Set(tips).size !== tips.length || tips.some((code, index) => code !== rows[index])) throw new Error('Species topology tips must uniquely match the stack-row order.');
    const valuesByCode = new Map(rows.map((code, index) => [
      code,
      typeof stack.subarray === 'function'
        ? stack.subarray(index * queryBases, (index + 1) * queryBases)
        : stack.slice(index * queryBases, (index + 1) * queryBases),
    ]));
    const species = rows.map((code, index) => summary('species', code, labels[index], [code], [valuesByCode.get(code)], chromosome, start, queryBases));
    const clades = cladeRecords(selectedTopology.tree).map(({ member_codes: memberCodes, ...clade }) => summary(
      'clade', clade.id, clade.name, memberCodes, memberCodes.map((code) => valuesByCode.get(code)), chromosome, start, queryBases, clade,
    ));
    return {
      schema_version: SCHEMA_VERSION, analysis_version: ANALYSIS_VERSION,
      coordinate_convention: COORDINATE_CONVENTION, window_size: WINDOW_SIZE,
      window_detection_threshold: WINDOW_DETECTION_THRESHOLD,
      zero_semantics: 'No detected CNEr interval; not measured 0% identity.',
      query: { chromosome, start, end, bases: queryBases }, species_count: rows.length,
      species, clades,
    };
  }

  function displayRows(source, { selectedCodes = null, order = 'topology', collapsedClades = [] } = {}) {
    const selected = new Set(selectedCodes || source.stackRows);
    const collapsed = new Set(collapsedClades);
    const labelByCode = new Map(source.stackRows.map((code, index) => [code, source.stackSpecies[index]]));
    const visit = (node, path = []) => {
      if (typeof node === 'string') return selected.has(node) ? [{ id: node, kind: 'species', name: labelByCode.get(node), memberCodes: [node] }] : [];
      const currentPath = [...path, node.name];
      const id = currentPath.join('/');
      const members = tipCodes(node).filter((code) => selected.has(code));
      if (!members.length) return [];
      if (collapsed.has(id)) return [{ id, kind: 'clade', name: `${node.name} (${members.length} spp.)`, memberCodes: members }];
      return node.children.flatMap((child) => visit(child, currentPath));
    };
    const rows = visit(source.stackTopology.tree);
    if (order === 'alphabetical') rows.sort((left, right) => left.name.localeCompare(right.name));
    else if (order !== 'topology') throw new Error('Species display order must be topology or alphabetical.');
    return rows;
  }

  function displayTree(node, rowIds, path = []) {
    const allowed = new Set(rowIds);
    const visit = (current, currentPath) => {
      if (typeof current === 'string') return allowed.has(current) ? current : null;
      const nextPath = [...currentPath, current.name];
      const id = nextPath.join('/');
      if (allowed.has(id)) return id;
      const children = current.children.map((child) => visit(child, nextPath)).filter(Boolean);
      if (!children.length) return null;
      return children.length === 1 ? children[0] : { name: current.name, children };
    };
    return visit(node, path);
  }

  function summarizeDisplayHeatmap(source, baseSummary, rows) {
    const width = source.end - source.start + 1;
    const rowIndex = new Map(source.stackRows.map((code, index) => [code, index]));
    const cells = rows.map((row) => baseSummary.bins.map((bin) => {
      let detected = 0; let identityTotal = 0;
      row.memberCodes.forEach((code) => {
        const offset = rowIndex.get(code) * width;
        bin.forEach((record) => {
          const value = Number(source.values.stack[offset + record.index]);
          if (Number.isFinite(value) && value !== 0) { detected += 1; identityTotal += value; }
        });
      });
      const possible = row.memberCodes.length * bin.length;
      return {
        identity: detected ? identityTotal / detected : 0,
        detectedFraction: detected / possible,
        genomicStart: Math.min(...bin.map((record) => record.position)),
        genomicEnd: Math.max(...bin.map((record) => record.position)),
      };
    }));
    return { ...baseSummary, cells };
  }

  return { analyzeSpeciesContext, cladeRecords, displayRows, displayTree, summarizeDisplayHeatmap, tipCodes };
}));
